### ISSUE TITLE
Segfault with PHP 7.0.2 & pthreads 3.1.5

### ISSUE THREAD

**Issue by JesusTheHun** [krakjoe/pthreads#549]
Hello,

How to install the case : https://gist.github.com/JesusTheHun/997edc8a47ac1155c79a
And this is the way to reproduce : 

Try to run :
`$ php app/console mt:test`

You'll get a fatal about a Closure. I was looking down for that Closure I couldn't find, tracking with xdebug what is actually happening.
Then I stepped on something. Open the file :

_vendor/symfony/symfony/src/Symfony/Component/Console/Application.php_

And line 856, between `$event = new ConsoleCommandEvent ...` and `$this->dispatcher->dispatch(ConsoleEvents::COMMAND, $event);`

Insert this code :

```
$w = new \Thread();
$w->start();
$w->join();
```

And now rerun :
`$ php app/console mt:test`

Segfault.
If you use a user class which defines run(), you'll see the segfault happens after the run() method is done :

```
$w = new class extends \Thread {
    public function run()
    {
        echo "Worker (".$this->getThreadId().") ::run";
    }
};
$w->start();
$w->join();
```

At this moment we just have our output `Worker (140590895331072) ::run`.
And then : 

`$this->dispatcher->dispatch(ConsoleEvents::COMMAND, $event);`

Segfault happens when we execute THIS LINE, which is weird, isn't it ?

I don't know how much the context and the object hierarchy can affect this...
If I can be of any help, just ask.

**Comment by krakjoe** [comment]
I've investigated here, there are other bugs, also raised by trying to load symfony ... the line the fault happens is not useful in either case, what is really happening is heap corruption, that has strange symptoms ...

Sorry I haven't got anything more useful to say right now ...

**Comment by shiro0x00** [comment]
Is there any progress? I have got the same bug with php 7.17 and Symfony 3.3

**Comment by tpunt** [comment]
@shiro0x00 Can you share your reproducing code for this? Ideally something that I can just clone and then run, since I couldn't setup OP's reproducing case.

**Comment by JesusTheHun** [comment]
If I remember correctly, the issue was the execution of a closure stored in an array :

`$callbacks = [function(){ echo "do something"; }]`

I don't have a php+pthread install right now to test it.

I started to work on this issue, atm it was in src/store.c, the method called to copy the context during the initiailization of a start didn't know how to deal with closures in array. Despite my efforts I never reached any working patch for this.

This is my last version : https://www.diffchecker.com/J3Z7ds5D, compared to what I think was the base version for work.

**Comment by shiro0x00** [comment]
Hi @tpunt ,
I created a github Repo: https://github.com/shiro0x00/SegFault
Use:
git clone https://github.com/shiro0x00/SegFault
cd SegFault
composer install
php bin/console app:test-thread

In the composer install process: When he asks for parameters just skip them. Skip the Errormessage. It is just an cache: clear error

Since the whole Symfony Framework is being downloaded it can take a while, even though I just use the Command Component.

I run this code under the latest raspbian, php7.1.7 (pthreads enabled), symfony 3.3

**Comment by tpunt** [comment]
Thanks, that worked. Though I could not reproduce the heap corruption issue mentioned above, and it didn't involve closures either. The segfault I received was related to sockets.

-- notes to my future self --

A `SIGPIPE` occurs due to a write attempt on a broken TCP connection. I made a quick fix to this in my copy of php-src ([offending line](https://github.com/php/php-src/blob/7bb696dc49efbe9354e6940c8d5a59105b29cc61/main/streams/xp_socket.c#L260)) by setting the `MSG_NOSIGNAL` flag. I'm not sure if this is the correct solution, though (I don't think PHP should segfault here, but given that the code in question has remained largely the same for 10+ years, it would suggest there is a deeper issue with pthreads instead).

Running the app with the temporary fix showed another segfault later on, this time on thread preparation (specifically when copying classes from a parent thread to a child thread). The parent thread's class table is `NULL` for some reason - I'm not quite sure when this can actually happen.

I'll try and take a deeper look into this soon.

**Comment by krakjoe** [comment]
As far as I can tell, this was fixed when we required 7.2 ... if I'm wrong, please re-open the issue.
